volatility3/framework/plugins/linux/lsblk.py

```diff
@@ -0,0 +1,447 @@
+                if not block_device:
+                    continue
+
+            try:
+                size = block_device.bd_inode.i_size
+            except exceptions.InvalidAddressException:
+                continue
+            # lsblk does not display devices with a size of 0
+            if not size:
+                continue
+
+            GENHD_FL_REMOVABLE = 1 << 0
+            removable = gendisk.flags & GENHD_FL_REMOVABLE
+
+            read_only = self._get_read_only(vmlinux, gendisk)
+
+            name = self._get_name(vmlinux, block_device, gendisk)
+
+            device_type = self._get_type(vmlinux, block_device, gendisk)
+
+            mountpoint = self._get_mountpoint(vmlinux, block_device)
+
+            yield 0, (
+                name,
+                self._major(device.devt),
+                self._minor(device.devt),
+                removable,
+                read_only,
+                self._format_bytes(size),
+                device_type,
+                mountpoint,
+            )
+
+    def _major(self, dev):
+        """Extract the major number from the device number."""
+        return (dev >> 20) & 0xFFF
+
+    def _minor(self, dev):
+        """Extract the minor number from the device number."""
+        return dev & 0xFF
+
+    def _get_read_only(self, vmlinux, gendisk):
+        # before v5.11 has a different check for read only that uses hd_struct
+        if not vmlinux.has_type("hd_struct"):
+            GD_READ_ONLY = 1
+            read_only = (
+                1
+                if (
+                    gendisk.part0.bd_read_only
+                    or (gendisk.state & (1 << GD_READ_ONLY)) != 0
+                )
+                else 0
+            )
+        else:
+            read_only = gendisk.part0.policy
+        return read_only
+
+    def _get_block_device(self, vmlinux, devt):
+        """In linux versions before 5.11, this function is how you get the block_device struct. Mimics the
+        function struct block_device* bdget(dev_t dev) https://elixir.bootlin.com/linux/v5.4/source/fs/block_dev.c#L900
+        """
+
+        blockdev_superblock = vmlinux.object_from_symbol("blockdev_superblock")
+        inode_hashtable = vmlinux.object_from_symbol("inode_hashtable")
+
+        # This the hash functions used to get the inode number https://elixir.bootlin.com/linux/v5.4/source/fs/block_dev.c#L867
+        val = self._major(devt) + self._minor(devt)
+
+        # Need these next 4 values for this hash function: https://elixir.bootlin.com/linux/v5.4/source/fs/inode.c#L474
+        i_hash_mask = vmlinux.object_from_symbol("i_hash_mask")
+        i_hash_shift = vmlinux.object_from_symbol("i_hash_shift")
+
+        # in linux 4.7, the GOLDEN_RATIO_PRIME constant changes. This struct gains a member from 4.6 to 4.7
+        if vmlinux.get_type("file_operations").has_member("iterate_shared"):
+            GOLDEN_RATIO_PRIME = 0x61C8864680B583EB
```

Comment:
Feels like these should be stored in the symbol table itself as a constant.  Are these not symbol values?  Such specific hard coded values are really bad for maintenance and things changing in the future...

Reply:
At the very least, this should live in `constants.linux` and have different names that can then be chosen by a plugin depending on which version the plugin decides to use.

Reply:
I added it to constants.linux